### PR TITLE
DM-41670: Account for non-atomic APDB writes in error-handling

### PR DESCRIPTION
**Pull request by kfindeisen**
This PR updates the specs of several `Apdb` methods to clarify their behavior regarding partial writes or database optimizations.

### CODE REVIEW

File: python/lsst/dax/apdb/apdb.py [hunk 1]
```diff
@@ -180,7 +180,8 @@ def tableDef(self, table: ApdbTables) -> Table | None:
     def getDiaObjects(self, region: Region) -> pandas.DataFrame:
         """Return catalog of DiaObject instances from a given region.
 
-        This method returns only the last version of each DiaObject. Some
+        This method returns only the last version of each DiaObject, and may
+        contain only a subset of the DiaObject columns. Some
```
Comment (ktlim) [inline python/lsst/dax/apdb/apdb.py:184]: I'd say something about the subset: it is the subset of columns required for AP association, and you get a subset unless the database has not been optimized.  As a result, this method is not suitable for DiaObject retrieval for general analysis purposes.

File: python/lsst/dax/apdb/apdb.py [hunk 2]
```diff
@@ -288,8 +290,8 @@ def containsVisitDetector(self, visit: int, detector: int) -> bool:
         Returns
         -------
         present : `bool`
-            `True` if some DiaObject, DiaSource, or DiaForcedSource records
-            exist for the specified observation, `False` otherwise.
+            `True` if at least one DiaSource or DiaForcedSource record
+            may exist for the specified observation, `False` otherwise.
```
Comment (ktlim) [inline python/lsst/dax/apdb/apdb.py:294]: Why the "may" here?  Could this return `True` if someone else is simultaneously removing all the `DiaSource` or `DiaForcedSource` records?  Might "at least one record exists at the time of checking" be a better phrasing?

**Review by ktlim — APPROVED** [review]

File: python/lsst/dax/apdb/apdb.py (hunk 2 above)
Comment (kfindeisen) [reply python/lsst/dax/apdb/apdb.py:294]: One proposed implementation on the Cassandra side is to have a separate table of visit-detector combinations that would be updated before writing sources [see Jira discussion](https://rubinobs.atlassian.net/browse/DM-41670?focusedCommentId=512188). It's possible for an error to occur after that table has been updated but before any actual sources have been written.

I think such false positives are acceptable (which is why I didn't complain), but the spec should reflect the fact that (for error handling purposes) it's pessimistic. There is precedent for this kind of ambiguity in the Stack, e.g. in [DatasetQueryResults.any](https://pipelines.lsst.io/v/daily/py-api/lsst.daf.butler.registry.queries.DatasetQueryResults.html#lsst.daf.butler.registry.queries.DatasetQueryResults.any).

File: python/lsst/dax/apdb/apdb.py (hunk 1 above)
Comment (kfindeisen) [reply python/lsst/dax/apdb/apdb.py:184]: I'll defer that to @andy-slac; while we didn't talk about this particular point, in our previous discussions, he tended to ask for specs to be as open-ended as possible.

File: python/lsst/dax/apdb/apdb.py (hunk 2 above)
Comment (ktlim) [reply python/lsst/dax/apdb/apdb.py:294]: So the utility of the method is that if it returns `False` you definitely know there is nothing to retrieve (yet), and if it returns `True`, you know that someone tried to write records, but you might or might not be able to retrieve any?

File: python/lsst/dax/apdb/apdb.py (hunk 1 above)
Comment (andy-slac) [reply python/lsst/dax/apdb/apdb.py:184]: `Apdb` interface is not meant for general analysis, only for AP pipeline use. But it's OK to say that subset must satisfy AP association needs, that is certainly what it has to do.

**Pull request merged**